storage/gcs.go

```diff
@@ -1,38 +1,161 @@
-	//TODO implement me
-	panic("implement me")
+	return UploadZip(ctx, resource, file, uploadFileGCS(g.client, g.bucket, nil))
 }
 
-func (g *gcs) UploadDir(ctx context.Context, resource Resource, source string) error {
-	//TODO implement me
-	panic("implement me")
+// UploadDir uploads the entire src directory to GCS.
+func (g *gcs) UploadDir(ctx context.Context, resource Resource, src string) error {
+	return UploadDir(ctx, resource, src, uploadFileGCS(g.client, g.bucket, resource))
 }
 
+// Download returns the URL to a zip file that contains all the contents of the given Resource.
 func (g *gcs) Download(ctx context.Context, resource Resource) (string, error) {
-	//TODO implement me
-	panic("implement me")
+	if err := validateResource(resource); err != nil {
+		return "", err
+	}
+
+	path := getZipLocation("", resource)
+	obj := getObjectGCS(g.client, g.bucket, path)
+	if _, err := obj.Attrs(ctx); err != nil {
+		if err == storage.ErrObjectNotExist {
+			return "", fmt.Errorf("object %s does not exist in bucket %s", path, g.bucket)
+		}
+		return "", err
+	}
+
+	opts := &storage.SignedURLOptions{
+		GoogleAccessID: g.accessID,
+		PrivateKey:     g.privateKey,
+		Scheme:         storage.SigningSchemeV4,
+		Method:         "GET",
+		Expires:        time.Now().Add(g.duration),
+	}
+
+	u, err := g.client.Bucket(g.bucket).SignedURL(path, opts)
+	if err != nil {
+		return "", err
+	}
+	return u, nil
 }
 
 // GetFile returns the content of file from a given path.
 func (g *gcs) GetFile(ctx context.Context, resource Resource, path string) ([]byte, error) {
-	//TODO implement me
-	panic("implement me")
+	return ReadFile(ctx, resource, path, readFileGCS(g.client, g.bucket))
+}
+
+// readFileS3v1 generates a function that contains the interaction with S3 to read the contents of a file.
+func readFileGCS(client *storage.Client, bucket string) ReadFileFunc {
+	return func(ctx context.Context, resource Resource, path string) (io.ReadCloser, error) {
+		obj := getObjectGCS(client, bucket, getLocation("", resource, path))
+		r, err := obj.NewReader(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return r, nil
+	}
+}
+
+func getObjectGCS(client *storage.Client, bucket string, path string) *storage.ObjectHandle {
+	obj := client.Bucket(bucket).Object(path)
+	return obj
+}
+
+// uploadFileS3v1 generates a function that uploads a single file in a path.
+func uploadFileGCS(client *storage.Client, bucket string, resource Resource) WalkDirFunc {
+	return func(ctx context.Context, path string, body io.Reader) error {
+		// If Resource is nil, it will use the given path as-is, otherwise it will use the given path as a relative path
+		// to the given Resource.
+		if resource != nil {
+			path = getLocation("", resource, path)
+		}
+
+		obj := getObjectGCS(client, bucket, path)
+		w := obj.NewWriter(ctx)
+
+		if _, err := io.Copy(w, body); err != nil {
+			return err
+		}
+
+		if err := w.Close(); err != nil {
+			return err
+		}
```

Comment:
This will never be called if `io.Copy` returns an error, which depending on the internal GCS implementation might be fine, but it can potentially leak memory.

I believe this code was taken from the GCS documentation, but what happens if we don't call `w.Close`? Should we maybe defer it and ignore the `Close()` error instead?